libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp

```diff
@@ -18,191 +17,100 @@
-            return false;
-        }
-    } else {
-        // something went wrong during the write
-        log_error() << "Failed to write to STDIN of dbus-proxy!";
-        return false;
-    }
+     log_debug() << "Activating dbus session bus";
+     ReturnCode sessionBusActivationResult;
+     try {
+         sessionBusActivationResult = sessionBus.activate();
+     } catch (GatewayError e) {
+         sessionBusActivationResult = ReturnCode::FAILURE;
+     }
+
+     log_debug() << "Activating dbus system bus";
+     ReturnCode systemBusActivationResult ;
+     try {
+         sessionBusActivationResult = systemBus.activate();
+     } catch (GatewayError e) {
+         systemBusActivationResult = ReturnCode::FAILURE;
+     }
+
+     if (isError(sessionBusActivationResult) && isError(systemBusActivationResult)) {
+         std::string err = "Neither dbus session bus nor dbus system bus could be activated";
+         log_error() << err;
+         throw GatewayError(err);
+     }
+     return ReturnCode::SUCCESS;
 }
 
-bool DBusGateway::startDBusProxy(const std::vector<std::string> &commandVec, const std::vector<std::string> &envVec)
+ReturnCode DBusGateway::teardown()
 {
-    // Spawn dbus-proxy with access to its stdin.
-    try {
-        Glib::spawn_async_with_pipes( "."
-                                    , commandVec
-                                    , envVec
-                                    , Glib::SPAWN_STDOUT_TO_DEV_NULL // Redirect stdout
-                                        | Glib::SPAWN_STDERR_TO_DEV_NULL // Redirect stderr
-                                        | Glib::SPAWN_SEARCH_PATH // Search $PATH
-                                        | Glib::SPAWN_DO_NOT_REAP_CHILD // Lets us do waitpid
-                                    , sigc::slot<void>() // child setup
-                                    , &m_pid
-                                    , &m_infp // stdin
-                                    );
-    } catch (const Glib::Error &ex) {
-        log_error() << "Failed to launch dbus-proxy";
-        return false;
-    }
-
-    log_debug() << "Started dbus-proxy: " << m_pid;
-
-    return true;
+     log_debug() << "Tearing down dbus session bus";
+     ReturnCode sessionBusTeardownResult;
+     try {
+         sessionBusTeardownResult = sessionBus.teardown();
+     } catch (GatewayError e) {
+         sessionBusTeardownResult = ReturnCode::FAILURE;
+     }
+
+     log_debug() << "Tearing down dbus system bus";
+     ReturnCode systemBusTeardownResult ;
+     try {
+         sessionBusTeardownResult = systemBus.teardown();
+     } catch (GatewayError e) {
+         systemBusTeardownResult = ReturnCode::FAILURE;
+     }
+
+     if (isError(sessionBusTeardownResult) && isError(systemBusTeardownResult)) {
+         std::string err = "Neither dbus session bus nor dbus system bus could be activated";
+         log_error() << err;
+         throw GatewayError(err);
+     }
+     return ReturnCode::SUCCESS;
```

Comment:
And the same goes here. Perhaps we can check the gateway state here also before calling teardown so that we know if it is actually needed? That would still work with the container pattern since all gateways (or sub-objects) implements that method.